skimage/registration/_lddmm.py

```diff
@@ -0,0 +1,2211 @@
+        >>> # Define images.
+        >>> # The reference_image is registered to the moving_image image
+        >>> # but both transformations are returned.
+        >>> #
+        >>> # reference_image is a binary ellipse with semi-radii 5 and 8 in
+        >>> # dimensions 0 and 1. The overall shape is (19, 25).
+        >>> # moving_image is a 30 degree rotation of reference_image.
+        >>> #
+        >>> reference_image = np.array([[(col-12)**2/8**2 + (row-9)**2/5**2
+        ... <= 1 for col in range(25)] for row in range(19)], float)
+        >>> moving_image = rotate(reference_image, 30)
+        >>> #
+        >>> # Register the reference_image to the moving_image,
+        >>> # then deform the reference_image and moving_image
+        >>> # to match the other.
+        >>> #
+        >>> lddmm_output = diffeomorphic_metric_mapping(
+        ... reference_image, moving_image, deformative_stepsize=0.5)
+        >>> #
+        >>> deformed_moving_image = map_coordinates(moving_image,
+        ... lddmm_output.moving_image_to_reference_image_transform)
+        >>> #
+        >>> deformed_reference_image = map_coordinates(reference_image,
+        ... lddmm_output.reference_image_to_moving_image_transform)
+
+    """
+
+    # Validate kwargs.
+    recognized_kwarg_keys = {
+        "multiscales",
+        "num_iterations",
+        "num_affine_only_iterations",
+        "num_rigid_affine_iterations",
+        "affine_stepsize",
+        "fixed_affine_scale",
+        "velocity_smooth_length",
+        "preconditioner_velocity_smooth_length",
+        "maximum_velocity_fields_update",
+        "num_timesteps",
+        "spatially_varying_contrast_map",
+        "contrast_iterations",
+        "sigma_contrast",
+        "contrast_smooth_length",
+        "sigma_matching",
+        "artifact_and_background_classification",
+        "sigma_artifact",
+        "sigma_background",
+        "artifact_prior",
+        "background_prior",
+        "initial_affine",
+        "initial_contrast_coefficients",
+        "initial_velocity_fields",
+        "map_coordinates_ify",
+    }
+    invalid_kwargs = list(
+        filter(
+            lambda kwarg_key: kwarg_key not in recognized_kwarg_keys,
+            kwargs.keys(),
+        )
+    )
+    if invalid_kwargs:
+        raise ValueError(
+            "One or more unexpected keyword arguments were encountered.\n"
+            f"Unrecognized keyword arguments: {invalid_kwargs}."
+        )
+    # Augment kwargs with None entries for unspecified arguments.
+    for kwarg_key in recognized_kwarg_keys:
+        if kwarg_key not in kwargs.keys():
+            kwargs[kwarg_key] = None
+
+    # Validate images and spacings.
+    # Images.
+    reference_image = _validate_ndarray(reference_image, dtype=float)
+    moving_image = _validate_ndarray(
+        moving_image, dtype=float, required_ndim=reference_image.ndim
+    )
+    # spacing.
+    reference_image_spacing = _validate_scalar_to_multi(
```

Comment:
do you need to call `_validate_scalar_to_multi` on the spacings both here and in the class above? Probably not?

Reply:
The class is meant to handle exactly 1 registration, the function wraps it in the multiscales handling, and part of that involves rescaling the images and correspondingly adjusting their resolutions (spacings).

Soo, it's necessary in the diffeomorphic_metric_mapping function and _could_ be left out of the class, but it's a very cheap operation and I like that the class works for a single registration as a stand-alone. As-is it doesn't expect any handling to be done upstream.